File: 1254-number-of-closed-islands/1254-number-of-closed-islands.cpp
class Solution {
    int rows, cols;
    vector<pair<int, int>> directions{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    
    bool dfs(int y, int x, vector<vector<bool>>& visited, vector<vector<int>>& grid){
        if (y < 0 || y >= rows || x < 0 || x >= cols) return false;
        if (grid[y][x] || visited[y][x]) return true;
        visited[y][x] = true;
        bool isClosed = true;
        for (auto& [dy, dx] : directions){
            int new_rows = y + dy, new_cols = x + dx;
            if (!dfs(new_rows, new_cols, visited, grid)) isClosed = false;
        }
        
        return isClosed;
    }
    
public:
    int closedIsland(vector<vector<int>>& grid) {
        rows = grid.size(), cols = grid[0].size();
        vector<vector<bool>> visited(rows, vector<bool>(cols, false));
        int ans = 0;
        
        for (int y = 0; y < rows; ++y){
            for (int x = 0; x < cols; ++x){
                if (grid[y][x] == 0 && !visited[y][x] && dfs(y, x, visited, grid)) ++ans;
            }
        }
        
        return ans;
    }
};

File: 1254-number-of-closed-islands/1254-number-of-closed-islands.py
class Solution:
    def dfs(self, y: int, x: int, visited: List[List[bool]], grid: List[List[int]]) -> bool:
        if x < 0 or x >= self.cols or y < 0 or y >= self.rows:
            return False
        
        if grid[y][x] or visited[y][x]:
            return True
        
        visited[y][x] = True
        isClosed = True
        
        for dy, dx in self.directions:
            new_rows, new_cols = y + dy, x + dx
            if not self.dfs(new_rows, new_cols, visited, grid):
                isClosed = False
        return isClosed
            
        
    def closedIsland(self, grid: List[List[int]]) -> int:
        self.rows, self.cols, ans = len(grid), len(grid[0]), 0
        self.directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        visited = [[False] * self.cols for _ in range(self.rows)]
        
        for y in range(self.rows):
            for x in range(self.cols):
                if grid[y][x] == 0 and not visited[y][x] and self.dfs(y, x, visited, grid):
                    ans += 1
        
        return ans
                    